### PR TITLE
Enable PyTorch 1.7 in conda CI

### LINKED ISSUE #3531: add support for 1.7 nightly
## 🚀 Feature
 
@ydcjeff created a CI with Conda to have tested with nighly 1.7 but in #3074 some tests were failing for 1.7 some let's move it as separate work

### Motivation

be continuously compatible even with new PT versions when they come

### ToDo

- [ ] uncomment CI and enable 1.7 testing
- [ ] fix test so 1.7 passes

### PR DESCRIPTION
**Pull request by ydcjeff**
## What does this PR do?

Fixes #3531

Python property support for PyTorch JIT PR (v1.7): https://github.com/pytorch/pytorch/pull/42390

Thanks to [ptrblck](https://github.com/ptrblck)

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
lets first finis the tests fixing here and later follow up in #3543

**Comment by Borda** [comment]
it seems that most of the TorchScript are failing
```
FAILED tests/models/test_cpu.py::test_lbfgs_cpu_model - AssertionError: This ...
FAILED tests/models/test_torchscript.py::test_torchscript_input_output[EvalModelTemplate]
FAILED tests/models/test_torchscript.py::test_torchscript_input_output[ParityModuleRNN]
FAILED tests/models/test_torchscript.py::test_torchscript_input_output[BasicGAN]
FAILED tests/models/test_torchscript.py::test_torchscript_retain_training_state
FAILED tests/models/test_torchscript.py::test_torchscript_properties[EvalModelTemplate]
FAILED tests/models/test_torchscript.py::test_torchscript_properties[ParityModuleRNN]
FAILED tests/models/test_torchscript.py::test_torchscript_properties[BasicGAN]
FAILED tests/models/test_torchscript.py::test_torchscript_save_load[EvalModelTemplate]
FAILED tests/models/test_torchscript.py::test_torchscript_save_load[ParityModuleRNN]
FAILED tests/models/test_torchscript.py::test_torchscript_save_load[BasicGAN]
```

**Comment by ydcjeff** [comment]
Seems like there is something update to the JIT.
- Using torch 1.7.0.dev20200915, test failed https://github.com/PyTorchLightning/pytorch-lightning/runs/1139697173
- Using torch 1.7.0.dev20200914, test passed https://github.com/PyTorchLightning/pytorch-lightning/runs/1139697179

But not sure the cause is from Lightning or PyTorch.

**Comment by Borda** [comment]
Maybe move this testing to some alternative ci/badge? @williamFalcon

**Comment by ydcjeff** [comment]
`to_torchscript` now works with PyTorch 1.7. This is done via `__ignored_properties__`, from https://github.com/pytorch/pytorch/pull/42390/files#diff-ff4f8670281cd1eb4e09329cc1dcb43b

So far I have ignored all the properties in `LightningModule` (some are from DeviceDtypeMixin) that would cause torchscript tests failing.

Question:
- What other properties should we ignore and not ignore? (since I am not sure), but we are only converting LightningModule to TorchScript, right?

TorchScript doesn't support `Union` type.

There is a PR #3295 doing similarly for 1.7 JIT support doing with `@torch.jit.ignore`.

For these kind of changes, would it be better to have separate ci/badge?

@williamFalcon @Borda @nateraw @awaelchli

**Review by Borda — APPROVED** [review]
LGTM

**Review by SkafteNicki — APPROVED** [review]
LGTM

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -51,9 +51,13 @@
...
 
-class LightningModule(
-    ABC, DeviceDtypeModuleMixin, GradInformation, ModelIO, ModelHooks, DataHooks, Module
-):
+class LightningModule(ABC, DeviceDtypeModuleMixin, GradInformation, ModelIO, ModelHooks, DataHooks, Module):
+    # Below is for property support of JIT in PyTorch 1.7
+    # since none of them is important when using JIT,
+    # we are going to ignore them.
+    # https://github.com/pytorch/pytorch/commit/e7d782e724c76bb0572023d52ee7438a40a7a262#diff-ff4f8670281cd1eb4e09329cc1dcb43b
+    __ignored_properties__ = ['datamodule', 'device', 'dtype', 'example_input_array', 'hparams', 'on_gpu']
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:59]: you can do this:
```
    __ignored_properties__ = DeviceDtypeModuleMixin.__ignored_properties__ + [
        "example_input_array",
        "datamodule",
        "on_gpu",
        "hparams",
    ]
```

Comment (ydcjeff) [reply pytorch_lightning/core/lightning.py:59]: Thank you for that @ananthsub

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -51,9 +51,14 @@
     XLA_AVAILABLE = True
 
 
-class LightningModule(
-    ABC, DeviceDtypeModuleMixin, GradInformation, ModelIO, ModelHooks, DataHooks, Module
-):
+class LightningModule(ABC, DeviceDtypeModuleMixin, GradInformation, ModelIO, ModelHooks, DataHooks, Module):
+    # Below is for property support of JIT in PyTorch 1.7
+    # since none of them is important when using JIT,
+    # we are going to ignore them.
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:57]:
```suggestion
    # since none of them is important when using JIT, we are going to ignore them.
```

**Pull request merged**